Realm.Shared/Realm.cs

```diff
@@ -357,55 +365,77 @@ private TableHandle GetTable(Schema.Object schema)
         /// <exception cref="RealmOutsideTransactionException">If you invoke this when there is no write Transaction active on the realm.</exception>
         public T CreateObject<T>() where T : RealmObject, new()
         {
+            RealmObject.Metadata metadata;
+            var ret = CreateObject(typeof(T).Name, out metadata);
+            if (typeof(T) != metadata.Schema.Type)
+                throw new ArgumentException($"The type {typeof(T).FullName} does not match the original type the schema was created for - {metadata.Schema.Type?.FullName}");
+
+            return (T)ret;
+        }
+
+        /// <summary>
+        /// Factory for a managed object in a realm. Only valid within a Write transaction.
+        /// </summary>
+        /// <returns>A dynamically-accessed Realm object.</returns>
+        /// <param name="className">The type of object to create as defined in the schema.</param>
+        /// <remarks>
+        /// If the realm instance has been created from an untyped schema (such as when migrating from an older version of a realm) the returned object will be purely dynamic.
+        /// If the realm has been created from a typed schema as is the default case when calling <code>Realm.GetInstance()</code> the returned object will be an instance of a user-defined class, as if created by <code>Realm.CreateObject&lt;T&gt;()</code>.
+        /// </remarks>
+        public dynamic CreateObject(string className)
+        {
+            RealmObject.Metadata ignored;
+            return CreateObject(className, out ignored);
+        }
+
+        private RealmObject CreateObject(string className, out RealmObject.Metadata metadata)
+        {
             if (!IsInTransaction)
                 throw new RealmOutsideTransactionException("Cannot create Realm object outside write transactions");
 
-            var objectType = typeof(T);
-            if (Config.ObjectClasses != null && !Config.ObjectClasses.Contains(objectType))
-                throw new ArgumentException($"The class {objectType.Name} is not in the limited set of classes for this realm");
-
-            var metadata = Metadata[typeof(T)];
+            if (!Metadata.TryGetValue(className, out metadata))
```

Comment:
I can't say I'm surprised that a O(1) dictionary lookup is faster than the O(ln(n)) lookups for the sorted structures :) Compared to the previous implementation, I think the performance difference will be based on the speed of `GetHashCode` for `Type` and `string` which should be negligible. Additionally I'd be surprised if an array lookup significantly outperforms the dictionary, but that might be worth benchmarking.